### PR TITLE
Return helpful message when appending an `Instruction` class

### LINKED ISSUE #3858: Appending an `Instruction` class instead of an instance gives an unhelpful error message
### What is the expected enhancement?

If a user forgets the `()` for their gate, they'll see the following error message:
```
>>> import qiskit as qk
>>> from qiskit.extensions.standard import *
>>> qr = qk.QuantumRegister(2)
>>> qc = qk.QuantumCircuit(qr)
>>> qc.append(CnotGate, qr[:], [])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/quantumcircuit.py", line 462, in append
    for (qarg, carg) in instruction.broadcast_arguments(expanded_qargs, expanded_cargs):
TypeError: broadcast_arguments() missing 1 required positional argument: 'cargs'
```

because `instruction` will be an `Instruction` subclass, and not an instance. (`broadcast_arguments` will be called as a static method on the class, instead of an instance method so `self` will be missing. Thus the `TypeError` for missing `cargs`)

In `QuantumCircuit.append`, we already check `if not isinstance(instruction, Instruction)` before checking `hasattr(instruction, 'to_instruction')`.

This should be restructured so that if `instruction` is not an `Instruction` instance, and does not have a `to_instruction`, we raise. (Optionally, we can check if `issubclass(instruction, Instruction)` and if so, suggest adding `()`).

### PR DESCRIPTION
**Pull request by EnriqueL8**
### Summary
Fixes #3858 

I'm not sure if this change needs a release note, happy to add one if needed.

### CODE REVIEW

**Comment by Cryoris** [comment]
Appending the type `HGate` is not supported you have to pass the instance `HGate()`. I'm not sure if accepting subtypes of `Instruction` would be a very consistent behaviour, since we would need to instantiate the class which we can only do if it takes no arguments in the initializer. While in principle possible for some gates, this would e.g. not work for `RYGate` which is a fairly basic gate.

Is there some particular use case you have in mind where supporting types instead of instances is required or simplifies things a lot?

**Comment by EnriqueL8** [comment]
Great point! I wasn't aware of that, I thought the function would still accept `HGate` as a type and raise accordingly. Wasn't necessary thinking of specific use case,  I was just trying to implement the flow described in the issue, and throw a nice error when a user passes a type instead of an instance as an argument.

File: test/python/circuit/test_instructions.py [hunk 1]
```diff
@@ -332,6 +332,11 @@ def test_modifying_copied_params_leaves_orig(self):
 
         self.assertEqual(inst.params, [0, 1, 2])
 
+    def test_instance_of_instruction(self):
+        qr = QuantumRegister(2)
+        qc = QuantumCircuit(qr)
+        self.assertRaises(CircuitError, qc.append(HGate, qr[:], []))
```
Comment (Cryoris) [inline test/python/circuit/test_instructions.py:338]: You can also use `assertRaises` as context manager, like so, this should fix the errors.
```suggestion
        with self.assertRaises(CircuitError):
            qc.append(HGate, qr[:], [])
```

File: qiskit/circuit/quantumcircuit.py [hunk 2]
```diff
@@ -532,9 +532,17 @@ def append(self, instruction, qargs=None, cargs=None):
...
+        Raises:
+            CircuitError: if object passed is a subclass of Instruction
+            CircuitError: if object passed is neither subclass nor an instance of Instruction
         """
-        # Convert input to Instruction
-        if not isinstance(instruction, Instruction) and hasattr(instruction, 'to_instruction'):
+        # Convert input to instruction
+        if not isinstance(instruction, Instruction) and not hasattr(instruction, 'to_instruction'):
+            if issubclass(instruction, Instruction):
+                raise CircuitError('Object is a subclass of Instruction, please add () to pass an instance of this object.')
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:542]:
```suggestion
                raise CircuitError('Object is a subclass of Instruction, please add () to '
                                   'pass an instance of this object.')
```

Comment (EnriqueL8) [reply qiskit/circuit/quantumcircuit.py]: Thanks!

File: test/python/circuit/test_instructions.py (hunk 1 above)
Comment (EnriqueL8) [reply test/python/circuit/test_instructions.py]: why does the previous version fail? Shouldn't they be equivalent?

File: qiskit/circuit/quantumcircuit.py [hunk 3]
```diff
@@ -557,9 +557,18 @@ def append(self, instruction, qargs=None, cargs=None):
...
         """
-        # Convert input to Instruction
-        if not isinstance(instruction, Instruction) and hasattr(instruction, 'to_instruction'):
+        # Convert input to instruction
+        if not isinstance(instruction, Instruction) and not hasattr(instruction, 'to_instruction'):
+            if issubclass(instruction, Instruction):
+                raise CircuitError('Object is a subclass of Instruction, please add () to '
+                                   'pass an instance of this object.')
+
+            raise CircuitError('Object is neither subclass nor an instance of Instruction.')
```
Comment (Cryoris) [inline qiskit/circuit/quantumcircuit.py:570]:
```suggestion
            raise CircuitError('Object to append must be an Instruction or '
                               'have a to_instruction() method.')
```

**Pull request merged**